python/cudf/cudf/core/index.py

```diff
@@ -2752,6 +2762,236 @@ def categories(self):
+    freq : numeric, default None
+        The length of each interval. Must be consistent
+        with the type of start and end
+    name : str, default None
+        Name of the resulting IntervalIndex.
+    closed : {"left", "right", "both", "neither"}, default "right"
+        Whether the intervals are closed on the left-side, right-side,
+        both or neither.
+
+    Returns
+    -------
+    IntervalIndex
+
+    Examples
+    --------
+    >>> import cudf
+    >>> import pandas as pd
+    >>> cudf.interval_range(start=0,end=5)
+    IntervalIndex([(0, 0], (1, 1], (2, 2], (3, 3], (4, 4], (5, 5]],
+    ...closed='right',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, freq=2,closed='left')
+    IntervalIndex([[0, 2), [2, 4), [4, 6), [6, 8), [8, 10)],
+    ...closed='left',dtype='interval')
+    >>> cudf.interval_range(start=0,end=10, periods=3,closed='left')
+    ...IntervalIndex([[0.0, 3.3333333333333335),
+            [3.3333333333333335, 6.666666666666667),
+            [6.666666666666667, 10.0)],
+            closed='left',
+            dtype='interval')
+    """
+    if freq and periods and start and end:
+        raise ValueError(
+            "Of the four parameters: start, end, periods, and "
+            "freq, exactly three must be specified"
+        )
+    if not isinstance(start or freq or end, int) and not isinstance(
+        start or freq or end, float
+    ):
+        raise NotImplementedError("Non-numeric values not yet supported")
+    elif periods and not freq:
+        # if statement for mypy to pass
+        if end is not None and start is not None:
+            # determine if periods are float or integer
+            quotient, remainder = divmod((end - start), periods)
+            if remainder:
+                step_input = (end - start) / periods
+                if (
+                    step_input % 1 == 0
+                    and type(start) == int
+                    and type(end) == int
+                ):
+                    freq_step = cudf.Scalar(step_input, dtype=int).device_value
+                else:
+                    freq_step = cudf.Scalar(
+                        (end - start) / periods
+                    ).device_value
+            else:
+                if (
+                    quotient % 1 == 0
+                    and type(start) == int
+                    and type(end) == int
+                ):
+                    freq_step = cudf.Scalar(quotient, dtype=int).device_value
+                else:
+                    freq_step = cudf.Scalar(quotient).device_value
+            if (
+                type(start) == int
+                and type(end) == int
+                and freq_step.dtype == int
+            ):
+                start = cudf.Scalar(start, dtype=int).device_value
+            else:
+                start = cudf.Scalar(start, dtype=float).device_value
+            bin_edges = sequence(size=periods + 1, init=start, step=freq_step,)
```

Comment:
This seems to be working correctly with the new parameters I added to the tests. It seems pretty tricky to try typecast a Device scalar once it's been created, which is why I tried to do the checks beforehand.